README
LendingClub version 0.01
========================

This perl module is intended to act as a loose API for interacting with the website lendingclub.com.

INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install
   
In Windows systems, use nmake or dmake instead.

DEPENDENCIES

This module requires these other modules and libraries:

  WWW::Mechanize
  Text::CSV
  XML::Simple
  HTML::Entities

COPYRIGHT AND LICENCE

This library is free software; you can redistribute it and/or
modify it under the same terms as Perl itself.

Copyright (C) 2010 by Shane P Connelly

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.10.1 or,
at your option, any later version of Perl 5 you may have available.




Changes
Revision history for Perl extension LendingClub.

0.01  Fri Jul  2 21:51:22 2010
	- original version; created by h2xs 1.23 with options
		-AXc -n LendingClub



lib/LendingClub/Note.pm
package LendingClub::Note;

use 5.010000;
use strict;
use warnings;
use WWW::Mechanize;
use LendingClub::Loan;
use LendingClub::Member;

our $VERSION = '0.01';

sub new
{
  my $package = shift;
  
  my $passed_params = shift;
  my %params = ();
  
  while ( my($key,$value) = each %{$passed_params} ) {
    if ( exists $params{$key} ) {
        $params{$key} = $value;
    }
    else {
        $params{$key} = $value;
    }
  }
  
  my $self = bless({}, $package);
  
  my $self->{'loan'} = LendingClub::Loan->new({lendingclub => $self, mech => $self->{'mech'}});
  my $self->{'member'} = LendingClub::Loan->new({lendingclub => $self, mech => $self->{'mech'}});
  
  return $self;
}

sub get_data
{
  my $self = shift;
  
  my $noteid = $self->{'noteid'};
  
  if (! $self->{'lendingclub'}->{'is_logged_in'})
  {
    $self->{'lendingclub'}->debug("Detected as not logged in.");
    $self->{'lendingclub'}->login();
  }
  
  $self->{'lendingclub'}->debug("Getting note data for note ID " . $noteid);
  
  my $mech = $self->{'mech'};
  $mech->get( "https://www.lendingclub.com/browse/loanDetail.action?loan_id=" . $noteid );
  my $htmlver = $mech->content( );
  
  my $note = {};
  #### TODO: SPLIT THIS OUT INTO SEPARATE SUB-MODULES (FOR LOAN, MEMBER, AND CREDIT HISTORY?)
  
  #### FIRST, GET THE LOAN DETAILS ####
  $self->{'loan'}->parse_loandetail($htmlver);
  
  #### THEN, GET THE MEMBER INFO ####
  $note->{'member'}->{'home_ownership'}->{'value'} = $1 if ($htmlver =~ m!<th\s*>\s*Home\s+Ownership\s*</th>\s*<td\s*>\s*(.*?)\s*</td>!is);
  $note->{'member'}->{'current_employer'}->{'value'} = $1 if ($htmlver =~ m!<th\s*>\s*Current\s+Employer\s*</th>\s*<td\s*>\s*(.*?)\s*</td>!is);
  $note->{'member'}->{'length_of_employment'}->{'value'} = $1 if ($htmlver =~ m!<th\s*>\s*Length\s+of\s+Employment\s*</th>\s*<td\s*>\s*(.*?)\s*</td>!is);
  $note->{'member'}->{'gross_income'}->{'value'} = $1 if ($htmlver =~ m!<th\s*>\s*Gross\s+Income\s*</th>\s*<td\s*>\s*(.*?)\s*</td>!is);
  $note->{'member'}->{'debt_to_income_pct'}->{'value'} = $1 if ($htmlver =~ m!<th\s*>\s*Debt-to-Income\s+\(DTI\)\s*</th>\s*<td\s*>\s*(.*?)\s*</td>!is);
  $note->{'member'}->{'location'}->{'value'} = $1 if ($htmlver =~ m!<th\s*>\s*Location\s*</th>\s*<td\s*>\s*(.*?)\s*</td>!is);
  $note->{'member'}->{'name'}->{'value'} = $1 if ($htmlver =~ m!>(.*?)\'s\s+Profile</td>!is);
  
  #### ADD VERIFIED INFORMATION ####
  $note->{'member'}->{'home_ownership'}->{'verified'} = $self->check_verified($note->{'member'}->{'home_ownership'}->{'value'});
  $note->{'member'}->{'current_employer'}->{'verified'} = $self->check_verified($note->{'member'}->{'current_employer'}->{'value'});
  $note->{'member'}->{'length_of_employment'}->{'verified'} = $self->check_verified($note->{'member'}->{'length_of_employment'}->{'value'});
  $note->{'member'}->{'gross_income'}->{'verified'} = $self->check_verified($note->{'member'}->{'gross_income'}->{'value'});
  $note->{'member'}->{'debt_to_income_pct'}->{'verified'} = $self->check_verified($note->{'member'}->{'debt_to_income_pct'}->{'value'});
  $note->{'member'}->{'location'}->{'verified'} = $self->check_verified($note->{'member'}->{'location'}->{'value'});
  
  #### THEN, GET THE CREDIT HISTORY ####
  $note->{'member'}->{'credit_score_range'}->{'value'} = $1 if ($htmlver =~ m!<th\s*>\s*Credit\s+Score\s+Range\s*</th>\s*<td\s*>\s*(.*?)\s*</td>!is);
  $note->{'member'}->{'earliest_credit_line'}->{'value'} = $1 if ($htmlver =~ m!<th\s*>\s*Earliest\s+Credit\s+Line\s*</th>\s*<td\s*>\s*(.*?)\s*</td>!is);
  $note->{'member'}->{'open_credit_lines'}->{'value'} = $1 if ($htmlver =~ m!<th\s*>\s*Open\s+Credit\s+Lines\s*</th>\s*<td\s*>\s*(.*?)\s*</td>!is);
  $note->{'member'}->{'total_credit_lines'}->{'value'} = $1 if ($htmlver =~ m!<th\s*>\s*Total\s+Credit\s+Lines\s*</th>\s*<td\s*>\s*(.*?)\s*</td>!is);
  $note->{'member'}->{'revolving_credit_balance'}->{'value'} = $1 if ($htmlver =~ m!<th\s*>\s*Revolving\s+Credit\s+Balance\s*</th>\s*<td\s*>\s*(.*?)\s*</td>!is);
  $note->{'member'}->{'revolving_credit_utilization_pct'}->{'value'} = $1 if ($htmlver =~ m!<th\s*>\s*Revolving\s+Line\s+Utilization\s*</th>\s*<td\s*>\s*(.*?)\s*</td>!is);
  $note->{'member'}->{'recent_inquiries'}->{'value'} = $1 if ($htmlver =~ m!<th\s*>\s*Inquiries\s+in\s+the\s+Last\s+6\s+Months\s*</th>\s*<td\s*>\s*(.*?)\s*</td>!is);
  $note->{'member'}->{'delinquent_accounts'}->{'value'} = $1 if ($htmlver =~ m!<th\s*>\s*Accounts\s+Now\s+Delinquent\s*</th>\s*<td\s*>\s*(.*?)\s*</td>!is);
  $note->{'member'}->{'delinquent_amount'}->{'value'} = $1 if ($htmlver =~ m!<th\s*>\s*Delinquent\s+Amount\s*</th>\s*<td\s*>\s*(.*?)\s*</td>!is);
  $note->{'member'}->{'recent_delinquencies'}->{'value'} = $1 if ($htmlver =~ m!<th\s*>\s*Delinquencies\s+\(Last\s+2\s+yrs\)\s*</th>\s*<td\s*>\s*(.*?)\s*</td>!is);
  $note->{'member'}->{'last_delinquency_months_elapsed'}->{'value'} = $1 if ($htmlver =~ m!<th\s*>\s*Months\s+Since\s+Last\s+Delinquency\s*</th>\s*<td\s*>\s*(.*?)\s*</td>!is);
  $note->{'member'}->{'public_records_count'}->{'value'} = $1 if ($htmlver =~ m!<th\s*>\s*Public\s+Records\s+On\s+File\s*</th>\s*<td\s*>\s*(.*?)\s*</td>!is);
  $note->{'member'}->{'last_public_record_months_elapsed'}->{'value'} = $1 if ($htmlver =~ m!<th\s*>\s*Months\s+Since\s+Last\s+Record\s*</th>\s*<td\s*>\s*(.*?)\s*</td>!is);
  
  #### PULL IN ANY INFORMATION FROM A PREVIOUS get_invested_notes() CALL ####
  if ($self->{'notes_invested_received_csv'})
  {
    my @rows = @{ $self->{'notes_invested_rows'} };
    foreach my $row (@rows)
    {
      if ($row->{'NoteId'} == $noteid)
      {
        $note->{'loan'}->{'is_invested'} = 1;
        $note->{'loan'}->{'amount_lent'} += $row->{'AmountLent'};
        $note->{'loan'}->{'remaining_principal'} += $row->{'PrincipalRemaining'};
        $note->{'loan'}->{'payments_received'} += $row->{'PaymentsReceivedToDate'};
        $note->{'loan'}->{'next_payment_date'} = $row->{'NextPaymentDate'};
        $note->{'loan'}->{'next_payment_date'} = 'N/A' if ($note->{'loan'}->{'next_payment_date'} eq '' || $note->{'loan'}->{'next_payment_date'} eq 'null');
        $note->{'loan'}->{'in_portfolio_names'} .= ($row->{'PortfolioName'} . ',');
        $note->{'loan'}->{'in_portfolio_ids'} .= ($row->{'PortfolioId'} . ',');
      }
    }
    if ($note->{'loan'}->{'is_invested'})
    {
      chop($note->{'loan'}->{'in_portfolio_ids'});
      chop($note->{'loan'}->{'in_portfolio_names'});
    }
  }
  
  #### PULL IN ANY INFORMATION FROM A PREVIOUS get_available_notes() CALL ####
  if ($self->{'notes_available_received_csv'})
  {
    my @rows = @{ $self->{'notes_available_rows'} };
    foreach my $row (@rows)
    {
      if ($row->{'Id'} == $noteid)
      {
        $note->{'member'}->{'job_title'} = $row->{'JobTitle'};
        $note->{'member'}->{'job_title'} = 'N/A' if ($note->{'member'}->{'job_title'} eq '' || $note->{'member'}->{'job_title'} eq 'null');
        $note->{'member'}->{'employment_status'} = $row->{'EmpStatus'};
        $note->{'member'}->{'employment_status'} = 'N/A' if ($note->{'member'}->{'employment_status'} eq '' || $note->{'member'}->{'employment_status'} eq 'null');
        $note->{'member'}->{'gross_income'}->{'verification_status'} = $row->{'IncomeVStatus'};
        $note->{'member'}->{'gross_income'}->{'verification_status'} = 'N/A' if ($row->{'IncomeVStatus'} eq '' || $row->{'IncomeVStatus'} eq 'null');
      }
    }
  }
  
  #### THEN, FORMAT EACH OF THESE PIECES OF INFORMATION ####
  
  if ($note->{'member'}->{'length_of_employment'}->{'value'} =~ m!\<!i)
  {
    $note->{'member'}->{'length_of_employment'}->{'value'} = 'N/A';
  }
  else
  {
    $note->{'member'}->{'length_of_employment'}->{'value'} = $self->format_numeric($note->{'member'}->{'length_of_employment'}->{'value'});
    if ($note->{'member'}->{'length_of_employment'}->{'value'} ne 'N/A')
    {
      $note->{'member'}->{'length_of_employment'}->{'value'} *= 12;
    }
  }
  
  $note->{'member'}->{'gross_income'}->{'value'} = $self->format_dollar_amount($note->{'member'}->{'gross_income'}->{'value'});
  $note->{'member'}->{'debt_to_income_pct'}->{'value'} = $self->format_percentage($note->{'member'}->{'debt_to_income_pct'}->{'value'});
  $note->{'member'}->{'open_credit_lines'}->{'value'} = $self->format_numeric($note->{'member'}->{'open_credit_lines'}->{'value'});
  $note->{'member'}->{'total_credit_lines'}->{'value'} = $self->format_numeric($note->{'member'}->{'total_credit_lines'}->{'value'});
  $note->{'member'}->{'revolving_credit_balance'}->{'value'} = $self->format_dollar_amount($note->{'member'}->{'revolving_credit_balance'}->{'value'});
  $note->{'member'}->{'revolving_credit_utilization_pct'}->{'value'} = $self->format_percentage($note->{'member'}->{'revolving_credit_utilization_pct'}->{'value'});
  $note->{'member'}->{'delinquent_amount'}->{'value'} = $self->format_dollar_amount($note->{'member'}->{'delinquent_amount'}->{'value'});
  
  $note->{'member'}->{'recent_inquiries'}->{'value'} = $self->format_numeric($note->{'member'}->{'recent_inquiries'}->{'value'});
  $note->{'member'}->{'delinquent_accounts'}->{'value'} = $self->format_numeric($note->{'member'}->{'delinquent_accounts'}->{'value'});
  $note->{'member'}->{'recent_delinquencies'}->{'value'} = $self->format_numeric($note->{'member'}->{'recent_delinquencies'}->{'value'});
  $note->{'member'}->{'last_delinquency_months_elapsed'}->{'value'} = $self->format_numeric($note->{'member'}->{'last_delinquency_months_elapsed'}->{'value'});
  $note->{'member'}->{'public_records_count'}->{'value'} = $self->format_numeric($note->{'member'}->{'public_records_count'}->{'value'});
  $note->{'member'}->{'last_public_record_months_elapsed'}->{'value'} = $self->format_numeric($note->{'member'}->{'last_public_record_months_elapsed'}->{'value'});
  
  return $note;
}

lib/LendingClub/Loan.pm
package LendingClub::Note;

use 5.010000;
use strict;
use warnings;
use WWW::Mechanize;

our $VERSION = '0.01';

sub new
{
  my $package = shift;
  
  my $passed_params = shift;
  my %params = ();
  
  while ( my($key,$value) = each %{$passed_params} ) {
    if ( exists $params{$key} ) {
        $params{$key} = $value;
    }
    else {
        $params{$key} = $value;
    }
  }
  
  my $self = bless({}, $package);
  
  return $self;
}

sub parse_loandetail
{
  my $self = shift;
  my $html = shift;
  $self->{'amount_requested'} = $1 if ($html =~ m!<th\s*>\s*Amount\s+Requested\s*</th>\s*<td\s*>\s*(.*?)\s*</td>!is);
  $self->{'loan_purpose'} = $1 if ($html =~ m!<th\s*>\s*Loan\s+Purpose\s*</th>\s*<td\s*>\s*(.*?)\s*</td>!is);
  $self->{'loan_grade'} = $1 if ($html =~ m!<th\s*>\s*Loan\s+Grade\s*</th>\s*<td[^\.]*?>\s*<span[^\.]*?>\s*(.*?)\s*</span>\s*</td>!is);
  $self->{'interest_rate'} = $1 if ($html =~ m!<th\s*>\s*Interest\s+Rate\s*</th>\s*<td\s*>\s*(.*?)\s*</td>!is);
  $self->{'monthly_payment'} = $1 if ($html =~ m!<th\s*>\s*Monthly\s+Payment\s*</th>\s*<td\s*>\s*(.*?)\s*</td>!is);
  $self->{'loan_length'} = $1 if ($html =~ m!<th\s*>\s*Loan\s+Length\s*</th>\s*<td\s*>\s*(.*?)\s*</td>!is);
  $self->{'funding_received'} = $1 if ($html =~ m!<th\s*>\s*Funding\s+Received\s*</th>\s*<td\s*>\s*(.*?)\s*</td>!is);
  $self->{'num_investors'} = $1 if ($html =~ m!<th\s*>\s*Investors\s*</th>\s*<td\s*>\s*(.*?)\s*</td>!is);
  $self->{'loan_status'} = $1 if ($html =~ m!<th\s*>\s*Loan\s+Status\s*</th>\s*<td\s*>\s*(.*?)\s*</td>!is);
  $self->{'loan_submitted_date_time'} = $1 if ($html =~ m!<th\s*>\s*Loan\s+Submitted\s+on\s*</th>\s*<td\s*>\s*(.*?)\s*</td>!is);
  $self->{'title'} = $1 if ($html =~ m!<h1>\s*(.*?)</h1>!is);
  $self->{'credit_review_status'} = $1 if ($html =~ m!<th\s*>\s*Credit\s+Review\s+Status\s*</th>\s*<td\s*>\s*(.*?)\s*</td>!is);
  $self->{'listing_expires_in'} = $1 if ($html =~ m!<th\s*>\s*Listing\s+Expires\s+In\s*</th>\s*<td\s*>\s*(.*?)\s*</td>!is);
  $self->{'description'} = $1 if ($html =~ m!<div\s+class="scrub"\s+id="loan_description">\s*(.*?)\s*</div>!is);
  
  if ($self->{'credit_review_status'})
  {
    $self->{'credit_review_status'} =~ s/<[^>]*?>//g;
    $self->{'credit_review_status'} =~ s/\&nbsp;//g;
    $self->{'credit_review_status'} =~ s/^\s*//g;
    $self->{'credit_review_status'} =~ s/\s*$//g;
  }
  
  if ($self->{'listing_expires_in'})
  {
    $self->{'listing_expires_in'} =~ s/\(.*?\)//;
    my $hr = $1 if ($self->{'listing_expires_in'} =~ m!(\d+)h!i);
    my $mn = $1 if ($self->{'listing_expires_in'} =~ m!(\d+)m!i);
    my $da = $1 if ($self->{'listing_expires_in'} =~ m!(\d+)d!i);
    if ($hr || $mn || $da)
    {
      $mn = 0 if (! $mn);
      $da = 0 if (! $da);
      $hr = 0 if (! $hr);
      $self->{'listing_expires_in'} = $mn + $hr * 60 + $da * 1440;
    }
    
  if ($self->{'funding_received'} =~ m!([\$\d\,\.]*)\s+\(([\%\d\,\.]*)\s+funded\)!)
  {
    $self->{'funding_received_amt'} = $self->{'lendingclub'}->format_dollar_amount($1);
    $self->{'funding_received_pct'} = $self->{'lendingclub'}->format_percentage($2);
  }
  }
  
  $self->{'description'} =~ s/<.*?>//g;
  
  $self->{'amount_requested'} = $self->{'lendingclub'}->format_dollar_amount($note->{'loan'}->{'amount_requested'});
  $self->{'interest_rate'} = $self->{'lendingclub'}->format_percentage($note->{'loan'}->{'interest_rate'});
  $self->{'monthly_payment'} = $self->{'lendingclub'}->format_dollar_amount($note->{'loan'}->{'monthly_payment'});
  $self->{'num_investors'} = $self->{'lendingclub'}->format_numeric($note->{'loan'}->{'num_investors'});
  if ($self->{'loan_length'} =~ m!(\d+)\s+year[s]*!i)
  {
    $self->{'loan_length'} = (int($1) * 12);
  }
  elsif ($self->{'loan_length'} =~ m!(\d+)\s+month[s]*!i)
  {
    $self->{'loan_length'} = int($1);
  }
  else
  {
    $self->{'loan_length'} = 'N/A';
  }
}

sub invest
{
  my $self = shift;
  my $amount = shift;
  my $portfolioname = shift;
  
  my $loanid = $self->{'id'};
  
  my $url = 'https://www.lendingclub.com/browse/addToPortfolio.action?loan_id=' . $loanid . '&loan_amount=' . $amount;
  
  my $mech = $self->{'mech'};
  $mech->get( $url );
  my $txtcontent = $mech->content( );
  $mech->get( 'https://www.lendingclub.com/portfolio/viewOrder.action' );
  $txtcontent = $mech->content( );
  $mech->get( 'https://www.lendingclub.com/portfolio/placeOrder.action' );
  $txtcontent = $mech->content( );
  
  my $i = 0;
  foreach my $f ( @{ $mech->forms() } )
  {
    my $input = $f->find_input( 'place-order-link1' , 'submit' );
    last if (defined($input));
    $i++;
  }
  $i--;
  
  if ($i > -1)
  {
    $mech->submit_form(
      form_number => $i
    );
    
    $txtcontent = $mech->content( );
    
    if ($portfolioname)
    {
      #this part fails ...
      my $j = 0;
      my $formnum = -1;
      my $portfolioexists = 0;
      foreach my $fm ( $mech->forms() )
      {
        foreach my $inp ( $fm->inputs )
        {
          print $inp->class . ' = ' . $inp->id . ' = ' . $inp->value . "\n";
          if ( lc($inp->type) eq 'option' && lc($inp->value) eq lc($portfolioname) )
          {
            $portfolioname = $inp->value;
            $portfolioexists = 1;
          }
          elsif ( lc($inp->type) eq 'option' && lc($inp->class) eq 'new_portfolio' )
          {
            $formnum = $j;
          }
        }
        $j++;
      }
      if ($formnum > -1)
      {
        $txtcontent = $mech->content( );
        if ($txtcontent =~ m!name="order_id"\s+value="(.*?)"!is)
        {
          my $orderid = $1;
          my $request = HTTP::Request->new(POST => 'https://www.lendingclub.com/data/portfolioManagement');
          $request->content_type('application/x-www-form-urlencoded');
          my $portfolionameencoded = $portfolioname;
          $portfolionameencoded =~ s/([^A-Za-z0-9])/sprintf("%%%02X", ord($1))/seg;
          my $post = 'lcportfolio_description=&record_id=1&lcportfolio_name=' . $portfolionameencoded . '&loan_id=' . $loanid . '&order_id=' . $orderid;
          $post .= ($portfolioexists ? '&method=addToLCPortfolio' : '&method=createLCPortfolio');
          $request->content($post);
          $mech->request($request);
        }
      }
    }
  }
}